feat: Support reference types in Union for operations

seebees opened the pull request:
Because reference types can have mutable state,
they need to have ValidState and Modifies
threaded through on the Dafny side.

Constructors for the service already supported this,
but this adds support for specific operations.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

texastony commented on TestModels/Union/src/SimpleResource.dfy:
```diff
@@ -0,0 +1,135 @@
+  import opened Types = SimpleUnionTypes
+  import opened Wrappers
+
+  /*
```
Praise: This doc is great.

texastony commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithydafny/DafnyApiCodegen.java:
```diff
@@ -1312,6 +1325,104 @@ private TokenTree OperationMemberValidState(
+          TokenTree.of("\n && (match %s ".formatted(varName)),
+          TokenTree.of(union
+            .members()
+            .stream()
+            .filter(this::OnlyReferenceStructures)
+            .map(s -> TokenTree.of(
```
Suggestion: I am not sure what the line comment should be, 
but there should be a line comment... 
Because Smithy-Dafny is complicated.
```suggestion
            .stream()
            // Reference Structures in a Union need Modifies
            .filter(this::OnlyReferenceStructures)
            .map(s -> TokenTree.of(
```

seebees replied:
added, but the code moved. Added a slightly different comment, but still something

texastony commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithydafny/DafnyApiCodegen.java:
```diff
@@ -1421,6 +1537,83 @@ private TokenTree OperationMemberModifies(
+      );
+    } else if (isUnion && !member.isRequired()) {
+      // Optional union item
+      // Required union item
```
```suggestion
```

texastony replied:
This is the optional case only,
you handled the required case above,
right?

seebees replied:
fixed

robin-aws commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/traits/MutableLocalStateTraitValidator.java:
```diff
@@ -0,0 +1,69 @@
+          """
+          aws.polymorph#mutableLocalState may allow callers to introduce soundness issues.
+
+          At this time to support dynamic mutable state smithy-dafny creates a seperated class by using an {:axiom}.
```
```suggestion
          At this time to support dynamic mutable state smithy-dafny creates a "separated class" by using an {:axiom}.
```

Since this doesn't yet appear in any public RFC (and if it would help I could post mine)

robin-aws commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/traits/MutableLocalStateTraitValidator.java:
````diff
@@ -0,0 +1,69 @@
+          assert state == concreteResource.state;
+          expect state != concreteResource.state;
+          ```
+          """
````
I would add the recommendation to use export sets on each concrete trait implementation, since that defeats this attack.

seebees replied:
Yes, this makes sense.
I also added all this to the TestModel to prove that it really will work.

robin-aws commented on TestModels/Resource/Model/resources.smithy:
```diff
@@ -83,3 +97,12 @@ structure SimpleResourcesException {
+@aws.polymorph#reference(resource: MutableResource)
+structure MutableResourceReference {}
+
+@smithy.api#suppress(["MutableLocalStateTrait"])
```
TIL about the @suppress trait, nice!

FYI you can customize the identifier of the validation event (either by changing the name of the validator or passing it explicitly to `danger(...)`. We could be more specific and make it `MutableLocalStateTraitUnsound` or something, but I don't feel strongly.

seebees replied:
I get it. I don't feel that strongly. So I'm going to leave it.

robin-aws commented on TestModels/Resource/test/SimpleResourcesTest.dfy:
```diff
@@ -73,5 +78,76 @@ module SimpleResourcesTest {
+    // and the Types file was correct
+    // This is a basic check to make sure
+    // that this simplified separated class works.
+    expect before + 1 == test.MyInternalState;
```
Great test model update overall, thanks!

robin-aws commented on TestModels/Union/src/SimpleResource.dfy:
```diff
@@ -0,0 +1,135 @@
+
+    method GetResourceData' ( input: GetResourceDataInput )
+      returns (output: Result<GetResourceDataOutput, Error>)
+      requires
```
Musing: boy is it unfortunate that you have you copy the spec here, especially when you don't have to for service operations (since module refinement doesn't make you copy the spec) :P

seebees replied:
I know! It has come up before that such things should just get inserted in the same way with traits...

robin-aws reviewed (changes requested):
Looks really solid, looks like a great UX improvement and I think we've contained the soundness issue as much as possible.

Make sure you update the commit message/description to also mention the change to `@mutableLocalState`.

seebees commented:
For merging. I plan on rebasing the commits into 2 commits with clean commit messages, and then merging with rebase and merge as the strategy.

robin-aws commented:
> For merging. I plan on rebasing the commits into 2 commits with clean commit messages, and then merging with rebase and merge as the strategy.

I usually squash and merge, do you have a strong preference for rebasing instead?

seebees commented:
>>For merging. I plan on rebasing the commits into 2 commits with clean commit messages, and then merging with rebase and merge as the strategy.

>I usually squash and merge, do you have a strong preference for rebasing instead?

For this? Yes. Because what I'm doing is mixing what would normally be 2 PRs into a single PR. So I want the long term history to reflect what I intend.

Once the CI passes, I'll come back, clean up and you can approve :)

robin-aws commented:
> > > For merging. I plan on rebasing the commits into 2 commits with clean commit messages, and then merging with rebase and merge as the strategy.
> 
> > I usually squash and merge, do you have a strong preference for rebasing instead?
> 
> For this? Yes. Because what I'm doing is mixing what would normally be 2 PRs into a single PR. So I want the long term history to reflect what I intend.
> 
> Once the CI passes, I'll come back, clean up and you can approve :)

Ah that's a decent reason, I'll let you merge as you like then. I don't personally care as much about separating logical changes if we end up with PRs that change more than one thing, but I appreciate it anyway.

seebees committed:
feat: Support reference types in Union for operations

Because reference types can have mutable state,
they need to have ValidState and Modifies
threaded through on the Dafny side.

Constructors for the service already supported this,
but this adds support for specific operations.

feat: Support reference types in Union for operations

Because reference types can have mutable state,
they need to have ValidState and Modifies
threaded through on the Dafny side.

Constructors for the service already supported this,
but this adds support for specific operations.

seebees committed:
feat: Change MutableLocalStateTrait

The original MutableLocalStateTrait
used pure dynamic frames.
But moving forward we are trying to move
to seperated classes.
These constructs would let a class
have internal state that is represented
by the reper/ValidState style invariant
but this mutation would not leak
into callers `modfies` clause.

At this time such a language feature does not exist.
So we aproximate it with an axiom.

working around https://github.com/smithy-lang/smithy-dafny/issues/646
